### PR TITLE
"Reverse" JSON instances

### PR DESCRIPTION
**Pull request by JackKelly-Bellroy**
I wanted to use HAL types to send a ProxyRequest to a lambda as part of a CodeDeploy pre-traffic hook. To do this, I need a ToJSON instance for ProxyRequest (and to check the result, I'll need a FromJSON instance for ProxyResponse).

This PR adds the instances I want. I'm going to use this branch in a `source-repository-package` stanza in `cabal.project` and test, but I'm putting it up as draft so that it can get feedback.

### CODE REVIEW

**Comment by JackKelly-Bellroy** [comment]
I would prefer to use `hedgehog` over `QuickCheck` for two reasons:

1. You get automatic shrinking when you construct a generator
2. You don't have to choose between adding "irrelevant" `build-depends` to a library or orphan instances in the test suite.

I have a feeling that there's going to be weird interactions between string hashing and case-insensitive keys that will actually make it impossible to construct a round-tripping pair of instances. But I want to do some test deploys first, and see if I am generating payloads that my other lambda functions will actually accept.

**Comment by IamfromSpace** [comment]
That sounds good, if your willing to set it up.  My only real reason to use QuickCheck is its ready to get going with HSpec, without magic.

if there’s an easier story around instances that’s definitely positive here.

> I have a feeling that there's going to be weird interactions between string hashing and case-insensitive keys that will actually make it impossible to construct a round-tripping pair of instances.

Agreed.  I think CI preserves original casing, so I _think_ it will work.  Hopefully there aren’t any insurmountable hurdles.

**Comment by JackKelly-Bellroy** [comment]
This is now done, but it's not possible to get watertight round-tripping:

* The generator for `ProxyResponse` never puts a `Content-Type` into `multiValueHeaders`
* The `instance FromJSON ProxyResponse` does not put the `Content-Type` header into `multiValueHeaders`; only into the `ProxyBody`. These two facts mean that the `contentType` field of `ProxyBody` will be written out and parsed correctly during tests.
  - It might be worth allowing the `contentType` field in `ProxyBody` to override any `Content-Type` header set in the `multiValueHeaders` of a `ProxyResponse` (in `instance ToJSON ProxyResponse`, use `H.insert` instead of `H.insertWith (\_ old -> old)`). I have not done this because it would technically be a breaking change.
* The `authJson` generator (used to populate the `authorizer` field in `RequestContext`) never generates a `null`. This is because `instance FromJSON (RequestContext a)` uses `(.:?)` to parse the authorizer field if it exists, and aeson parses a present key/`null` value as `Nothing`, and also parses an absent field into `Nothing`. So if we generate `Just Null`, we fail to roundtrip.

File: src/AWS/Lambda/Events/ApiGateway/ProxyResponse.hs [hunk 1]
```diff
@@ -88,6 +93,14 @@ import           Network.HTTP.Types.Status (Status (..), accepted202,
                                             unprocessableEntity422,
                                             unsupportedMediaType415,
                                             upgradeRequired426, useProxy305)
+import           GHC.Generics               (Generic (..))
+
+-- This function is available in Data.Functor as of base 4.11, but we define it
+-- here for compatibility.
+(<&>) :: Functor f => f a -> (a -> b) -> f b
+(<&>) x f = f <$> x
```
Comment (JackKelly-Bellroy) [inline src/AWS/Lambda/Events/ApiGateway/ProxyResponse.hs:101]: Could we raise the `base` lower bound? `base-4.11` is GHC 8.4.1 (Mar 2018).

**Comment by JackKelly-Bellroy** [comment]
I have some deploys I want to test this with before I push for release but this is ready for a proper review now.

**Comment by JackKelly-Bellroy** [comment]
The pre-traffic hooks at work (which is where I wanted to use the instances) all seem to be working, so I think this is ready to be considered for merging.

File: src/AWS/Lambda/Events/ApiGateway/ProxyResponse.hs (hunk 1 above)
Comment (IamfromSpace) [reply src/AWS/Lambda/Events/ApiGateway/ProxyResponse.hs:101]: I like that this lib is very broadly compatible.  Certainly at some point it'll make sense to start dropping support of older versions--but this seems like a pretty minor convenience here to do so :)

**Comment by IamfromSpace** [comment]
This looks great!  I really appreciate you taking the time to add the property tests here; I was clearly overly optimistic about the level of effort here!  TY for putting in the extra mile :)

I kicked off the build check and it looks like there's a few failures.  They're configured to shortcut, so we can only see the first to fail.  Given the new manual changes by Github I should consider changing that in general...  I'll see if I can get them all to run so we can see all the issues at once.

**Comment by JackKelly-Bellroy** [comment]
@IamfromSpace All right, give this a go. I had to split the `stack.yaml` files because the `extra-deps` section gets pretty wild for old LTS resolvers.

Dropping GHC 8.2.2 would also save us the `import Data.Semigroup ((<>))` everywhere. How many people are using a compiler from 2017?

At this point, I suppose it doesn't matter since the branch should completely pass CI now.

**Review by IamfromSpace — APPROVED** [review]

**Comment by JackKelly-Bellroy** [comment]
> I’ll be able to release this late this evening.

Thanks very much. One question, before you cut the release: are you happy with the violations of round-tripping in https://github.com/Nike-Inc/hal/pull/90#issuecomment-832369688 ? Just wondering whether we should change the instance behaviour or document it it some way before release.

**Comment by IamfromSpace** [comment]
Ah, yes, good call.  I meant to write back on these.  I think I’m happy with these as invariants.  The actual authorizer payload is actually a JSON subset I believe, from [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-output.html):

> The output can also include a context map containing key-value pairs.

So it seems like we already handle more than we could even encounter.

For the Content-Type, for better or worse (still could be convince either way, haha), the idea was that it’s an innate part of the body, and should only be set as such.  I’d though about overrides, but dismissed it in the end.  So it too would be an invariant that it would never be populated in the headers themselves.

So I think this is all acceptable—but let me know if there’s anything else worth considering!

**Comment by JackKelly-Bellroy** [comment]
If you're happy with the tripping, then I am too. This is the only thing I'd consider changing:

https://github.com/Nike-Inc/hal/pull/90/files#diff-1d6ffc7a65b309a357c44cab1c8dde611b96bebeaf30bc49663c515331d47f31R216-R220

If the content-type is part of the body, then its setting should override any content-type set in `multiValueHeaders`. But it's an edge case and technically a breaking change. I don't think it will matter much, in practice.

**Pull request merged**